### ISSUE TITLE
Core flags given after task names fail silently

### ISSUE THREAD

**Issue by bitprophet** [pyinvoke/invoke#466]
Thanks to #205 we can now do things like `inv mytask --help` to get per-task help.

Unfortunately, a bug snuck in (_quelle horreur!_), namely that other core flags given after task names seem to just...do nothing. For example, `inv --echo mytask` shows command echoing, but `inv mytask --echo` does not. No error, but also no echo.

At first I feared this meant that _any_ post-task flags that were invalid were getting swallowed silently, but this doesn't seem to be the case, things like `inv mytask --lolno` still result in parser errors.

So I think this is probably just a TODO that isn't implemented yet, where we need to add more logic/behavior interpreting the rest of the core flags on a per-task-execution basis.

**Comment by bitprophet** [comment]
Figured this out just now while poking a related issue; it's simply because the code added for #205 did so in a very backwards compatible fashion (added `self.core_via_tasks` instead of directly/globally modifying `self.args`) and then only the "print help/list tasks" part of the code (`Program.parse_cleanup`) was updated to look at `self.core_via_tasks`.

The rest of the core arg attending code (i.e. that which sets things like `echo`; it's in `Program.update_config`) was not updated and still only refers to `self.args` - thus why these other core flags disappear into the aether.

Have a fix in place and as usual updating the tests is gonna take a bunch more time than the fix itself 😓

**Comment by bitprophet** [comment]
Also, wondering if this is truly a bug or an expectations mismatch:

- With `--help`, the intent was that `inv taskname -h` implicitly acts like `inv -h taskname` -  where the task name is itself used as the value for `-h`
- I have skipped tests in place that state the other core flags like `--list` should _actually not work_ in "task contexts", which goes against this ticket here and possibly the docs (need to check).
- Those together imply this was only ever really supposed _to_ work for `--help`, which is not what I keep remembering (eg this bug was filed because I keep habitually trying to slap `-e` at the end of command lines, etc).
- Typically, my gut instinct should be the guide in such situations (if I assume as much, others will too) but I need to see what our docs currently say; if we're really explicit about it and there's a stated reason, maybe we need to go the other way here (delineate `--help` as being special after all, and explode with parse errors for anything else). I'd rather not tho.

**Comment by bitprophet** [comment]
The docs appear very specific that this behavior was intended to be for `--help`; nowhere does it state that it "should" work for any/all core flags. Makes me wonder if anybody else is even noticing this or if it's uniquely a "Jeff problem"; seems likely the latter.

Of course, given it's easily fixable, I still think I'd like to do so, I just gotta tidy up that test suite and enhance the docs.

**Comment by bitprophet** [comment]
I suppose another way to look at the "you can slap `-e` and friends after tasks" is that it's purely a convenience/extra/gimme, in the situations where it's unambiguous. Users should be taught that these mostly belong in the core parse context, and that _if_ they happen to work outside of it, that's a happy coincidence – but one not to be relied upon.

Another argument is simply that there's no putting the `inv taskname -h` genie back in the bottle, and so this is all merely a matter of degree. Plus, having it work for `-h` but not `-e` is (as seen by me filing this ticket to begin with) is its own inconsistency.

There's no clear-cut winning here, so I am thinking we might as well roll with the convenience, as much as it pains the purity-minded side of me.

**Comment by bitprophet** [comment]
Diving into this and mostly done, I find that yanno...the split between `Program.args` and `Program.core_via_tasks` really is dumb and all over the place. I think old-me was in fact being too paranoid when he made `core_via_tasks` separate.

Given that this parser-exposed behavior was not even possible beforehand, and we're technically changing behavior (albeit from `ParseError` to non-error behavior), trying not to change the overall value of `.args` isn't going to actually help anybody subclassing/hacking these classes, avoid at least some pain or surprise.

**Comment by bitprophet** [comment]
Went halfsies and decided to make this overall a feature-level-bugfix; this way it limits patch level stability issues.

Seem to have actually sorted out behavior & tests (including a potential roadblock in the parser, which was fixed via some `#showerthoughts`) so this might be done now (!).

**Comment by bitprophet** [comment]
Ran into some extremely stupid sidequests and still not done with them. The main issue consists of the pseudo-internal APIs in Program and Executor; needed a way to easily represent the union of the "true" core flags (as parsed at the front of the CLI) and the "via tasks" core flags (the new behavior added here).

Initially decided not to modify `Program.core[0]` (a `ParserContext`, which is inside a `ParseResult`) but to instead add `Program.core_via_tasks` and then update `Program.args` (a property) so it reflects the union of both (as needed above). This took a while to nail down because there are some finicky edge cases.

Got that all working, only to find out that, oh. This works fine for core Invoke, but for downstream code like Fabric, which does work in a custom Executor that wants access to core flags (specifically in this case, a `--hosts` flag), the Program->Executor interface is too limited. Program creates Executor and hands it the loaded task collection, the loaded config, and `self.core` (the core args). (Then hands that instance the parsed task contexts to execute.)

The problem: the new shiny `Program.args` is thus invisible to `Executor` and it cannot see the unified core arg values, only the 'regular', given-up-front ones.

----

I'm glad I'm rubber ducking this because I think the solution might actually be easy. I was hesitating to go the obvious route - mutate `Program.core[0]` after all, in a single step after that final parse - because it felt backwards incompatible.

Then I remembered, well, older versions are arguably broken as-is - users can give the core args after tasks, and nothing happens (or, in even older versions - IIRC pre-1.0 - an explosion occurs because those flags are invalid in that position). Going from a silent error to the data becoming updated isn't a problem. It'd only _be_ a problem if there had been a prior release where this functionality already worked _and_ was implemented in a way that left `Program.core` unmolested. But no such release exists - it's just in my branch as of today.

Still means I have to shuffle all the new code and tests I wrote today, but, whatever. Wasted so much damn time on this as it is...
